<a href="https://colab.research.google.com/github/shemaaali/DS-Unit-1-Sprint-2-Statistics/blob/master/LS_DS_122_Chi2_Tests/S_A6_LS_DS18_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [120]:
# YOUR WORK HERE

import pandas as pd
from scipy import stats
import numpy as np
from scipy.stats import ttest_ind

In [2]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-07-14 18:32:59--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.03s   

2020-07-14 18:32:59 (699 KB/s) - ‘student.zip’ saved [20478/20478]



In [3]:
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [4]:
df = pd.read_csv('student-mat.csv', sep=';')

print(df.shape)
df.head()

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

In [5]:
df.describe(exclude='number')

school  sex address famsize  ... nursery higher internet romantic
count     395  395     395     395  ...     395    395      395      395
unique      2    2       2       2  ...       2      2        2        2
top        GP    F       U     GT3  ...     yes    yes      yes       no
freq      349  208     307     281  ...     314    375      329      263

[4 rows x 17 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [121]:
# YOUR WORK HERE
# Dependent variables
contingency = pd.crosstab(df['famrel'], df['freetime'])

contingency


freetime  1   2   3   4   5
famrel                     
1         1   3   3   1   0
2         1   3   7   5   2
3         6  17  24  16   5
4         9  26  83  55  22
5         2  15  40  38  11

In [82]:
contingency.values

array([[ 1,  3,  3,  1,  0],
       [ 1,  3,  7,  5,  2],
       [ 6, 17, 24, 16,  5],
       [ 9, 26, 83, 55, 22],
       [ 2, 15, 40, 38, 11]])

In [122]:
# contingency table with margins
contingency = pd.crosstab(df['famrel'], df['freetime'], margins=True)

contingency

freetime   1   2    3    4   5  All
famrel                             
1          1   3    3    1   0    8
2          1   3    7    5   2   18
3          6  17   24   16   5   68
4          9  26   83   55  22  195
5          2  15   40   38  11  106
All       19  64  157  115  40  395

In [123]:
# Independent variables
contingency = pd.crosstab(df['guardian'], df['freetime'])

contingency

freetime   1   2    3   4   5
guardian                     
father     4  18   33  25  10
mother    13  42  116  77  25
other      2   4    8  13   5

In [124]:
# contingency (observed) table with margins

observed = pd.crosstab(df['guardian'], df['freetime'], margins=True)

observed

freetime   1   2    3    4   5  All
guardian                           
father     4  18   33   25  10   90
mother    13  42  116   77  25  273
other      2   4    8   13   5   32
All       19  64  157  115  40  395

In [125]:
observed = observed.values
observed

array([[  4,  18,  33,  25,  10,  90],
       [ 13,  42, 116,  77,  25, 273],
       [  2,   4,   8,  13,   5,  32],
       [ 19,  64, 157, 115,  40, 395]])

In [126]:
# By hand Calculate the expected rows and columns:
# Calculate the expected rows and columns: (row_total)(column_total) / sample_size 

expected_row1_col1 = 90*19/395


expected_row1_col1

4.329113924050633

In [127]:
# Row totals for freetime
row_totals = df['freetime'].value_counts().sort_index().values

row_totals

array([ 19,  64, 157, 115,  40])

In [128]:
# column totals for guardian
col_totals = df['guardian'].value_counts().sort_index().values

col_totals

array([ 90, 273,  32])

In [129]:
sample_size = df.shape[0]

sample_size

395

In [130]:
expected = np.array([[0.0,0.0,0.0,0.0,0.0,0.0],
                     [0.0,0.0,0.0,0.0,0.0,0.0],
                     [0.0,0.0,0.0,0.0,0.0,0.0],
                     [0.0,0.0,0.0,0.0,0.0,0.0]])

for row_position, col in enumerate(col_totals):
  for column_position, row in enumerate(row_totals):
    expected_value = (row*col) / sample_size
    expected[row_position][column_position] = expected_value

expected

array([[  4.32911392,  14.58227848,  35.7721519 ,  26.20253165,
          9.11392405,   0.        ],
       [ 13.13164557,  44.23291139, 108.50886076,  79.48101266,
         27.64556962,   0.        ],
       [  1.53924051,   5.18481013,  12.71898734,   9.3164557 ,
          3.24050633,   0.        ],
       [  0.        ,   0.        ,   0.        ,   0.        ,
          0.        ,   0.        ]])

In [131]:
# Calcalute the chi^2 test 
chi2 = ((observed - expected)**2 / expected).sum().sum()

chi2

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: divide by zero encountered in true_divide
  


inf

In [132]:
chi2, p_value, dof, expected = stats.chi2_contingency(observed)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 6.7152901479558595
p value 0.9650740378661792
degrees of freedom 15
expected frequencies table 
 [[  4.32911392  14.58227848  35.7721519   26.20253165   9.11392405
   90.        ]
 [ 13.13164557  44.23291139 108.50886076  79.48101266  27.64556962
  273.        ]
 [  1.53924051   5.18481013  12.71898734   9.3164557    3.24050633
   32.        ]
 [ 19.          64.         157.         115.          40.
  395.        ]]


Conclusion:

Due to a chi^2 statistic of 6.715 and a p-value of 0.965, I fail to reject  the null hypothesis that freetime and guardian are independent.

In [133]:
contingency = pd.crosstab(df['famrel'], df['freetime'])

contingency

freetime  1   2   3   4   5
famrel                     
1         1   3   3   1   0
2         1   3   7   5   2
3         6  17  24  16   5
4         9  26  83  55  22
5         2  15  40  38  11

In [134]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 17.598623169878643
p value 0.34791854709653824
degrees of freedom 16
expected frequencies table 
 [[ 0.38481013  1.29620253  3.17974684  2.32911392  0.81012658]
 [ 0.86582278  2.9164557   7.15443038  5.24050633  1.82278481]
 [ 3.27088608 11.01772152 27.0278481  19.79746835  6.88607595]
 [ 9.37974684 31.59493671 77.50632911 56.7721519  19.74683544]
 [ 5.09873418 17.17468354 42.13164557 30.86075949 10.73417722]]


Conclusion:

Due to a chi^2 statistic of 17.59 and a p-value of 0.347, I fail to reject the null hypothesis that famrel and freetime are dependent.

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [ ]:
# YOUR WORK HERE

import numpy as np

In [97]:
observed

array([[  4,  18,  33,  25,  10,  90],
       [ 13,  42, 116,  77,  25, 273],
       [  2,   4,   8,  13,   5,  32],
       [ 19,  64, 157, 115,  40, 395]])

In [98]:
expected 

array([[  4.32911392,  14.58227848,  35.7721519 ,  26.20253165,
          9.11392405,   0.        ],
       [ 13.13164557,  44.23291139, 108.50886076,  79.48101266,
         27.64556962,   0.        ],
       [  1.53924051,   5.18481013,  12.71898734,   9.3164557 ,
          3.24050633,   0.        ],
       [  0.        ,   0.        ,   0.        ,   0.        ,
          0.        ,   0.        ]])

In [99]:
individual_chi2 = (observed - expected)**2 / expected
individual_chi2

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: RuntimeWarning: divide by zero encountered in true_divide
  """Entry point for launching an IPython kernel.


array([[2.50203562e-02, 8.01028482e-01, 2.14827058e-01, 5.51886508e-02,
        8.61462730e-02,            inf],
       [1.31975509e-03, 1.12719084e-01, 5.17166679e-01, 7.74452113e-02,
        2.53170353e-01,            inf],
       [1.37924714e-01, 2.70747628e-01, 1.75083447e+00, 1.45640134e+00,
        9.55350078e-01,            inf],
       [           inf,            inf,            inf,            inf,
                   inf,            inf]])

In [108]:
first_chi2 = (4.32911392-2.50203562e-02)**2 / 4.32911392

first_chi2

4.279217814149042

In [104]:
individual_chi2 = ((observed - expected)**2 / expected).sum().sum()
individual_chi2

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: RuntimeWarning: divide by zero encountered in true_divide
  """Entry point for launching an IPython kernel.


inf

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [109]:
# YOUR WORK HERE
observed = np.array([20,50,30,45,28,61,100])

observed

array([ 20,  50,  30,  45,  28,  61, 100])

In [110]:
expected = np.array([21.16, 21.16, 21.16, 21.16, 21.16, 21.16])

expected

array([21.16, 21.16, 21.16, 21.16, 21.16, 21.16])

In [111]:
# For cell 3 (index position 2)

observed[2] - expected[2]

8.84

In [112]:
# See when we squared term makes all the values positive and emphasizes places,
# where we saw a large deviation between observed and expected frequencies.
observed[2] - expected[2]**2 / expected

array([8.84, 8.84, 8.84, 8.84, 8.84, 8.84])

In [113]:
chi2 = ((observed[2] - expected[2])**2 / expected).sum()

chi2

22.158487712665405

In [117]:
from scipy.stats import normaltest
sample = np.random.poisson(5, 100)
print(normaltest(sample))  

NormaltestResult(statistic=0.6628797203563538, pvalue=0.717889328654845)




### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)